Documentation/rfc/v3api.md

```diff
@@ -0,0 +1,459 @@
+
+
+## Protobuf Defined API
+
+``` protobuf
+syntax = "proto3";
+
+// Interface exported by the server.
+service etcd {
+  // Range gets the keys in the range from the store.
+  // Maybe rename it to GET?
+  rpc Range(RangeRequest) returns (RangeResponse) {}
+
+  // Put puts the given key into the store.
+  // A put request increases the index of the store,
+  // and generates one event in the event history.
+  rpc Put(PutRequest) returns (PutResponse) {}
+
+  // Delete deletes the given key from the store.
+  // A delete request increase the index of the store,
+  // and generates one event in the event history.
+  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+
+  // Tnx processes all the requests in one transaction.
+  // A tnx request increases the index of the store,
+  // and generates one event in the event history.
+  rpc Tnx(TnxRequest) returns (TnxResponse) {}
+
+  // Watch watches the events happening or happened in etcd. Both input and output
+  // are stream. One watch rpc can watch for multiple ranges and get a stream of
+  // events. The whole events history can be watched unless compacted.
+  rpc Watch(stream WatchRequest) returns (stream WatchResponse) {}
+
+  // Compact compacts the event history in etcd. User should compact the
+  // event history periodically, or it will grow infinitely.
+  rpc Compact(CompactRequest) returns (CompactResponse) {}
+
+  // LeaseCreate creates a lease. A lease has a TTL. The lease will expire if the
+  // server does not receive a keepAlive within TTL from the lease holder.
+  // All keys attached to the lease will be expired and deleted if the lease expires.
+  // The key expiration generates an event in event history.
+  rpc LeaseCreate(LeaseCreateRequest) returns (LeaseCreateResponse) {}
+
+  // LeaseRevoke revokes a lease. All the key attached to the lease will be expired and deleted.
+  rpc LeaseRevoke(LeaseRevokeRequest) returns (LeaseRevokeResponse) {}
+
+  // LeaseAttach attaches keys with a lease.
+  rpc LeaseAttach(LeaseAttachRequest) returns (LeaseAttachResponse) {}
+
+  // LeaseTnx likes Tnx. It does one additional thing that all the keys put into the
+  // store are attached with the given lease. This is rpc is useful when you want to
+  // put a key and attach it to a lease atomically.
+  rpc LeaseTnx(LeaseTnxRequest) returns (LeaseTnxResponse) {}
+
+  // KeepAlive keeps the lease alive.
+  rpc LeaseKeepAlive(stream LeaseKeepAliveRequest) returns (stream LeaseKeepAliveResponse) {}
+}
+
+message RequestHeader {
+  // the cluster id to check with
+  // if the cluster id is not matched, an error is returned.
+  optional uint64 cluster_id = 1;
+}
+
+message ResponseHeader {
+  // an error type message?
+  optional string error = 1;
+  optional uint64 cluster_id = 2;
+  optional uint64 member_id = 3;
+  // index of the store when the requested was processed.
+  optional int64 index = 4;
+  optional uint64 raft_term = 5;
+}
+
+message RangeRequest {
```

Comment:
----- Original Message -----
> > +  // the cluster id to check with
> > +  // if the cluster id is not matched, an error is returned.
> > +  optional uint64 cluster_id = 1;
> > +}
> > +
> > +message ResponseHeader {
> > +  // an error type message?
> > +  optional string error = 1;
> > +  optional uint64 cluster_id = 2;
> > +  optional uint64 member_id = 3;
> > +  // index of the store when the requested was processed.
> > +  optional int64 index = 4;
> > +  optional uint64 raft_term = 5;
> > +}
> > +
> > +message RangeRequest {
> 
> @smarterclayton
> Let us be more clear here.
> 
> ```
> foo = bar         1
> foo = bar00     2: range query all -> foo = bar00; range query (all,
> max_index=1) -> nothing
> foo1 = bar       3: range query all -> foo = bar00; foo1 = bar 1; range query
> (all, max_index=2) -> foo = bar
> ```
> 
> Is this what you want? We are not going to keep all versions in the store.
> You can get the old versions only by scanning the history.
> 
> You can range scan the history via watch.
> 

Ok, that wasn't clear to me - the commit log will be in the store, but not the older versions?  I was working from the older set of assumptions where we having the older versions of keys be accessible.  From a use case perspective, here's some of my thoughts on older history:

1. We would like to be able to see the history for a particular key (current version -> older version 1 -> older version 2)
2. It's likely we would want to be able to see all of the deleted keys under a prefix until such a time as the store is compacted.

From what you described, 1 is available, but 2 would not be available unless a client choose to persist those older names in another key space?